sys/net/gnrc/conn/tcp/gnrc_conn_tcp.c

```diff
@@ -0,0 +1,560 @@
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+int conn_tcp_create(conn_tcp_t *conn, const void *addr, size_t addr_len, int family, uint16_t port)
+{
+    conn->l4_type = GNRC_NETTYPE_TCP;
+
+    /* Address Initialization */
+    switch (family) {
+#ifdef MODULE_GNRC_IPV6
+        case AF_INET6:
+            if (addr_len != sizeof(ipv6_addr_t)) {
+                return -EINVAL;
+            }
+            if (gnrc_conn6_set_local_addr(conn->local_addr, addr)) {
+                conn->l3_type = GNRC_NETTYPE_IPV6;
+
+                ipv6_addr_set_unspecified((ipv6_addr_t *) &(conn->peer_addr));
+                conn->local_addr_len = addr_len;
+                conn->peer_addr_len  = addr_len;
+            }
+            else {
+                return -EADDRNOTAVAIL;
+            }
+            break;
+#endif
+        default:
+            (void)addr;
+            (void)addr_len;
+            (void)port;
+            return -EAFNOSUPPORT;
+    }
+    /* Port Initialization */
+    conn->local_port = port;
+    conn->peer_port = GNRC_TCP_PORT_UNSPEC;
+
+    /* Initialize Transmission Control Block variables */
+    conn->state = CLOSED;
+    conn->snd_una = 0;
+    conn->snd_nxt = 0;
+    conn->snd_wnd = 0;
+    conn->snd_wl1 = 0;
+    conn->snd_wl2 = 0;
+    conn->rcv_nxt = 0;
+    conn->rcv_wnd = 0;
+    conn->iss = 0;
+    conn->irs = 0;
+    conn->mss = 0;
+    conn->status = 0;
+    conn->rtt_start = 0;
+    conn->rtt_var = GNRC_TCP_RTO_UNINITIALIZED;
+    conn->srtt = GNRC_TCP_RTO_UNINITIALIZED;
+    conn->rto = GNRC_TCP_RTO_UNINITIALIZED;
+    conn->retries = 0;
+    conn->rcv_buf_raw = NULL;
+    conn->pkt_retransmit = NULL;
+    mutex_init(&(conn->fsm_lock));
+    conn->owner = 0;
+    return 0;
+}
+
+int conn_tcp_close(conn_tcp_t *conn)
+{
+    msg_t msg;
+    msg_t msg_tout;
+    xtimer_t usr_tout;
+    int err = 0;
+
+    assert(conn->l4_type == GNRC_NETTYPE_TCP);
+
+    /* Guard: Return is already closed */
+    if(conn->state == CLOSED){
+        return 0;
+    }
+
+    /* Guard: Caller must be the connections owner */
+    if (conn->owner != thread_getpid()) {
+        return -EPERM;
+    }
+
+    /* Guard: Only a opened connection can be closed */
+    if (conn->state != LISTEN && conn->state != SYN_RCVD
+        && conn->state != ESTABLISHED && conn->state != CLOSE_WAIT) {
+        return -EALREADY;
+    }
+
+    /* If the connection is passive and the handle is not accepted */
+    if (conn->status & GNRC_TCP_STATUS_PASSIVE && !(conn->status & GNRC_TCP_STATUS_ACCEPTED)) {
+        return -EBADF;
+    }
+
+    /* Purge message queue */
+    while(msg_try_receive(&msg) >= 0){
+    }
+
+    /* Setup Userspace Timeout */
+    msg_tout.type = MSG_TYPE_USER_TIMEOUT;
+    xtimer_set_msg(&usr_tout, GNRC_TCP_USER_TIMEOUT_DURATION, &msg_tout, thread_getpid());
+
+    /* Start closing sequence */
+    _fsm(conn, CALL_CLOSE, NULL, NULL, 0);
+
+    /* Wait until the connection has been closed */
+    while (conn->state != CLOSED) {
```

Comment:
This must be changed to:

    /* Wait until the connection has been closed */
    gnrc_tcp_fsm_state_t final_state = (conn->status & GNRC_TCP_STATUS_PASSIVE) ? LISTEN : CLOSED;
    while (conn->state != final_state) {
        msg_receive(&msg);
        switch (msg.type) {
            case MSG_TYPE_NOTIFY_USER:
                DEBUG("gnrc_conn_tcp : conn_tcp_close() : MSG_TYPE_NOTIFY_USER\n");
                break;

            default:
                DEBUG("gnrc_conn_tcp : conn_tcp_close() : other message type\n");
        }
    }